### PR TITLE
Add Pulp 3 container image built from PyPi

### PR DESCRIPTION
**Pull request by ehelms**
In our discussions, the idea was for this to live in the Pulp main repository and to at least start off as an all-in-one image. There is related work to add running via the containers to the Ansible installer (https://github.com/pulp/ansible-pulp3/pull/45). There are still some gaps to address here in terms of process:

 1) How to CI the image build itself
 2) How to auto-build the image with PyPi released assets
 3) Creation of the registry project the image will live on
 4) Testing the resulting image in a deployment / Pulp smash scenario

I'm sure there are some other elements to address but this gets us started.

This was inspired by my POC project [carafe](https://github.com/ehelms/carafe).

### CODE REVIEW

File: container/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+RUN scl enable rh-python36 'pip install gunicorn'
+RUN scl enable rh-python36 'pip install pulpcore'
+
+RUN scl enable rh-python36 'pip install $PLUGINS'
+
+COPY container-assets/settings.py /etc/pulp/settings.py
+RUN echo "SECRET_KEY = '`cat /dev/urandom | tr -dc 'a-z0-9!@#$%^&*(\-_=+)' | head -c 50`'" >> /etc/pulp/settings.py
+
+RUN scl enable rh-python36 "pulp-manager makemigrations pulp_app"
```
[the remark this answers is not in the document]
Comment (ehelms) [reply container/Dockerfile]: That is the team's plan, to add these closer to GA.

File: container/container-assets/migrate_pulp.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+scl enable rh-python36 "pulp-manager migrate --noinput auth"
```
[the remark this answers is not in the document]
Comment (ehelms) [reply container/container-assets/migrate_pulp.sh]: Just following the Pulp docs.

File: container/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+RUN scl enable rh-python36 "pulp-manager collectstatic --noinput"
+
+COPY container-assets/wait_on_postgres.py /usr/bin/wait_on_postgres.py
+COPY container-assets/wait_on_database_migrations.sh /usr/bin/wait_on_database_migrations.sh
+COPY container-assets/migrate_pulp.sh /usr/bin/migrate_pulp.sh
+COPY container-assets/pulp-core /usr/bin/pulp-core
+COPY container-assets/pulp-worker /usr/bin/pulp-worker
+COPY container-assets/pulp-resource-manager /usr/bin/pulp-resource-manager
+
+RUN echo "scl enable rh-python36 \"pulp-manager migrate --noinput $PLUGINS\"" >> /usr/bin/migrate_pulp.sh
```
[the remark this answers is not in the document]
Comment (ehelms) [reply container/Dockerfile]: Just following the Pulp docs for now.

File: container/container-assets/migrate_pulp.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (asmacdo) [reply container/container-assets/migrate_pulp.sh]: This is to work around a problem with Django. "The issue is that an app without migrations that depends on an app with migrations most likely won't be handled correctly. Apps without migrations are created in the database before apps with migrations."
https://github.com/pinax/django-user-accounts/issues/179#issuecomment-157876649

This step will hopefully be unnecessary after we commit the migrations.

**Comment by ehelms** [comment]
@bmbouter Sounded like from mailing list your request would be to split this into an image per service? I'm good with that direction, wanting clarity before I do work towards it.

**Comment by bmbouter** [comment]
@ehelms If you are able to split it into one image per service, I think that would be great!

**Comment by ehelms** [comment]
@bmbouter This is ready for another round of review with split, individual images based on a pulp-core image.

File: containers/images/pulp-core/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+ENV LANG=en_US.UTF-8
+ENV LANGUAGE=en_US.UTF-8
+ENV LC_ALL=en_US.UTF-8
+ENV PYTHONUNBUFFERED=0
+ENV DJANGO_SETTINGS_MODULE=pulpcore.app.settings
+
+RUN mkdir -p /etc/pulp
+
+RUN scl enable rh-python36 'pip install gunicorn'
```
Comment (evgeni) [inline containers/images/pulp-core/Dockerfile]: should these three be compacted to one `RUN` to save a few layers?

File: containers/images/pulp-core/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+RUN scl enable rh-python36 'pip install pulpcore'
+RUN scl enable rh-python36 'pip install $PLUGINS'
+
+COPY container-assets/settings.py /etc/pulp/settings.py
+RUN echo "SECRET_KEY = '`cat /dev/urandom | tr -dc 'a-z0-9!@#$%^&*(\-_=+)' | head -c 50`'" >> /etc/pulp/settings.py
+
+RUN scl enable rh-python36 "pulp-manager makemigrations pulp_app"
```
Comment (evgeni) [inline containers/images/pulp-core/Dockerfile]: should these three be compacted to one `RUN` to save a few layers?

File: containers/push.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+---
+- hosts: localhost
+  vars:
+    container_cli: 'docker'
+  vars_files:
+    - vars/defaults.yaml
+  tasks:
+    - name: 'Check for podman'
```
Comment (evgeni) [inline containers/push.yaml]: given both playbooks have "check for podman" and "set_fact" steps, should this be moved to a "select_container_cli" role?

**Comment by dralley** [comment]
@ehelms status?

**Comment by bmbouter** [comment]
This repository has moved to https://github.com/pulp/pulpcore. As such this PR needs to close and be reopened over there. Could you please:

1. Fork https://github.com/pulp/pulpcore on github
2. Open a new PR against https://github.com/pulp/pulpcore/compare?expand=1

If you're need any help doing this please visit in #pulp-dev on IRC for help. Thank you and sorry for the disruption.

**Comment by cognifloyd** [comment]
https://github.com/pulp/pulpcore/pull/13

**Pull request closed without merging**